File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import deque

def bfs(visited, graph, start):
    count = 1
    q = deque()
    q.append(start)
    visited[start] = True
    
    while q:
        now_node = q.popleft()
        
        for next_node in graph[now_node]:
            if not visited[next_node]:
                visited[next_node] = True
                q.append(next_node)
                count += 1
                
    return count


def solution(n, wires):
    answer = 101
    
    for i in range(len(wires)):
        graph = [[] for _ in range(n+1)]
        visited = [False] * (n+1)
        diff = []
        
        for j in range(len(wires)):
            if i == j:
                continue
            s, e = wires[j][0], wires[j][1]
            graph[s].append(e)
            graph[e].append(s)
        
        for k in range(1, n+1):
            if not visited[k]:
                diff.append(bfs(visited, graph, k))
                
        answer = min(answer, abs(diff[0]-diff[1]))        

    return answer

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] wires) {
        int answer = n;
        
        for(int q = 0; q < wires.length; q++){
            // 초기화
            ArrayList<ArrayList<Integer>> list = new ArrayList<>();
            for(int i = 0; i <= n; i++){
                list.add(new ArrayList<>());
            }
            boolean[] visited = new boolean[n+1];
            
            for(int j = 0; j < wires.length; j++){
                if(q == j){
                    continue;
                }
                
                // 값 부여
                int x = wires[j][0];
                int y = wires[j][1];

                list.get(x).add(y);
                list.get(y).add(x);
                
            }
                        
            // bfs
            int result = bfs(1, list, visited);
            int diff = Math.abs(result - (n-result));            
                
            answer = Math.min(answer, diff);
        }
        

        return answer;
    }
    
    public int bfs(int start, ArrayList<ArrayList<Integer>> list, boolean[] visited){
        int count = 1;
        Queue<Integer> queue = new LinkedList<>();
        
        queue.add(start);
        visited[start] = true;
        
        while(queue.size()>0){
            int now = queue.poll();
            
            for(int next: list.get(now)){
                if(!visited[next]){
                    queue.add(next);
                    count+=1;
                    visited[next] =true;
                }
            }
        }

        return count;
    }
}